narwhal/test-utils/src/lib.rs

```diff
@@ -454,6 +454,85 @@ pub fn make_certificates(
     rounds_of_certificates(range, initial_parents, keys, failure_probability, generator)
 }
 
+// Creates certificates for the provided rounds but also having slow nodes.
+// `range`: the rounds for which we intend to create the certificates for
+// `initial_parents`: the parents to use when start creating the certificates
+// `keys`: the authorities for which it will create certificates for
+// `slow_nodes`: the authorities which are considered slow. Being a slow authority means that we will
+//  still create certificates for them on each round, but no other authority from higher round will refer
+// to those certificates. The number (by stake) of slow_nodes can not be > f , as otherwise no valid graph will be
+// produced.
+pub fn make_certificates_with_slow_nodes(
+    committee: &Committee,
+    range: RangeInclusive<Round>,
+    initial_parents: Vec<Certificate>,
+    keys: &[PublicKey],
+    slow_nodes: &[(PublicKey, f64)],
+) -> (VecDeque<Certificate>, Vec<Certificate>) {
+    // ensure provided slow nodes do not account > f
+    let slow_nodes_stake: Stake = slow_nodes
+        .iter()
+        .map(|(key, _)| committee.authorities.get(key).unwrap().stake)
+        .sum();
+
+    assert_eq!(keys.len(), 4);
+    assert!(slow_nodes_stake < committee.validity_threshold());
+
+    let mut certificates = VecDeque::new();
+    let mut parents = initial_parents;
+    let mut next_parents = Vec::new();
+
+    for round in range {
+        next_parents.clear();
+        for name in keys {
+            let this_cert_parents =
+                this_cert_parents_with_slow_nodes(name, parents.clone(), slow_nodes);
+            let (_, certificate) =
+                mock_certificate(committee, name.clone(), round, this_cert_parents);
+            certificates.push_back(certificate.clone());
+            next_parents.push(certificate);
+        }
+        parents = next_parents.clone();
+    }
+    (certificates, next_parents)
+}
+
+// Returns the parents that should be used as part of a newly created certificate.
```

Comment:
Instead of specific to slow nodes, maybe name this more genericly, e.g. `select_parent_subset()` or something?

Reply:
Yeah I was mostly tried to remain domain-driven on this thus did stick with the "slow node". Let me see how I could better rename it.